[Bug 1307271] Move to structured logging

imbstack opened the pull request:
This removes our direct dependencies on a bunch of services and replaces it with mozlog-compatible logging.

Bugzilla Bug: [1307271](https://bugzilla.mozilla.org/show_bug.cgi?id=1307271)

jhford commented on libraries/monitor/src/logger.js:
```diff
@@ -0,0 +1,119 @@
+ * We can consider supporting extra logging standards
+ * later if we want.
+ */
+class Logger {
```
Any chance we could have child loggers?  I suspect being able to pass in a set of fields to the constructor and returning a new Logger instance from a child method could achieve this?

imbstack replied:
Child loggers are supported at the lib-monitor level via `prefix(<name>, {extra-stuff})`. Each child ends up with its own `Logger`. I just did it that way since that is the way we supported children before so it will be easier to port.

jhford replied:
cool!

jhford commented on libraries/monitor/src/timekeeper.js:
```diff
@@ -0,0 +1,37 @@
+   * Compare the start and end times then submit the value to the monitor for
+   * records.
+   */
+  measure(force = false) {
```
Can we add the ability to pass in extra fields here? to include in the `.info` message?

imbstack commented:
I've updated for one of the things John asked for and gotten the rest of the tests passing. I figure that we can do a first round where we just replace the things that `monitor` did before and then come by in a second pass and convert a lot of `debug` stuff to this.

TODO:

- [ ] Update README and docs for monitor
- [ ] Figure out how to stuff log events into sentry, etc

imbstack commented:
My preferred way of us handling this would be: 

````js
const logger = require('./main')('monitor').prefix('whatever');
logger.debug('hello');
```

That way we always get all of our config from the same place and we can make sure that multiple instances of logging don't need to be kept in-sync. I've made #228 to allow for this to work in a nice way.

If that doesn't work in a certain situation, it should be decently easy to require lib-monitor directly and create a new logger. It will just require a bit of care to keep things  like serviceName in sync.

Does that make sense to you? I'm not married to any particular style.

jhford commented:
That's a neat idea, but I'm pretty sure that it won't work since the loader is async, and module level async operations aren't supported (at least to my knowledge).  You're right that some values need to be synchronised between the module logger and the logger from the loader, but if the root logger/monitor was the module level one, and the logging component in the loader just pulled in a `.prefix()`'d monitor instance, there wouldn't be any synchronising needed.

Since the root logger shouldn't need any/much configuration, the configuration function would likely be simple enough to have at the top of the module:

```javascript
const log = require('taskcluster-lib-monitor')('taskcluster-service-1');
```

then in the main.js file, have a component like:

```javascript
{ requires: [],
  setup: async () => {
    log.prefix(....);
  }
}
```

imbstack commented:
Ah yes. One solution to the async thing is in the other bug where I changed lib-loader but your solution works even better. So long as the general plan makes sense I'm happy.

djmitche commented on services/notify/src/main.js:
```diff
@@ -4,11 +4,11 @@ const {Client, pulseCredentials} = require('taskcluster-lib-pulse');
 const docs = require('taskcluster-lib-docs');
 const taskcluster = require('taskcluster-client');
 const _ = require('lodash');
+const Monitor = require('./monitor');
```
Maybe `const monitorBuilder` or something like that?

djmitche commented on services/notify/src/api.js:
```diff
@@ -31,6 +31,7 @@ builder.declare({
   ].join('\n'),
 }, async function(req, res) {
   debug(`Received request to send email to ${req.body.address}`);
+  this.monitor.email({address: req.body.address});
```
Purely on the basis of readability, I'd prefer this to be something like `this.monitor.log.email(..)`.  Maybe it would read better with a past-tense name like `this.monitor.emailSent`?

imbstack commented:
I now consider this largely ready for final review. Open question and TODOs:

- [ ] show a way for us to replace sentry/statsum
- [ ] consider switching the name of the service in lib-monitor from `projectName` to the shorter `serviceName`
- [ ] Consider using schemas to define the fields of specific types of messages and assert they are correct in development only
- [ ] your thing here? Let me know!

imbstack commented on generated/references.json:
```diff
@@ -9497,6 +9809,18 @@
         "/references/queue/v1/exchanges.json",
         "/references/secrets/v1/api.json",
         "/references/treeherder/v1/exchanges.json",
+        "/references/undefined/v1/logs.json",
```
Fixed! Good catch.

imbstack commented on libraries/api/test/responsetimer_test.js:
```diff
@@ -47,49 +47,42 @@ suite('api/responsetimer', function() {
       projectName: 'tc-lib-api-test',
-      credentials: {clientId: 'fake', accessToken: 'fake'},
+    });
+    monitorBuilder.setup({
```
I don't think it should be too hard to make that change. I don't mind much either way so I'll switch to `Manager`!

imbstack commented:
My plan with invalid messages would be to crash in development and probably not even validate in production. If we do end up validating in production we can just log an extra line that says, "invalid message" or something along those lines. Another option would be to log a single message that says, "invalid message" but with the original attachd as a field?

jhford commented:
I think crashing in development and not validating in production is a great idea.  That should catch the cases where the values are structurally invalid, but allow for cases where the values are falsy when they shouldn't be.  That's often a really good indication of a problem when logging -- something is undefined when it ought to be not-undefined.

For those messages which have structure defined, I wonder if we could set a default of a string `UNSET_PROPERTY` for all of the required properties, so that when the real properties come in through the logging function call, all of the provided required properties overwrite the placeholder properties.  In that case, it would be really easy to find cases in production of things not logging where we want them by searching for `UNSET_PROPERTY`.  This is probably too much, though.

imbstack commented:
I've opted to just do a simple assertion that any fields you've defined in the docs are present in the object you try to log in development. Nothing fancier than that. Maybe in a followup we can get fancier with schemas and such.

imbstack commented on generated/docs/notify/references/logs.json:
```diff
@@ -0,0 +1,116 @@
+      "description": "A request to send an irc message.",
+      "fields": {
+        "user": "A user, if this is a request to direct message a user",
+        "channel": "A channel (including #), if this is a request to post a message to a channel"
```
Yeah, this is the specific case that made me want a schema. I'm going to simplify this message for now and we can complicate it again if we move to schema or whatever.

imbstack commented on libraries/monitor/README.md:
```diff
@@ -10,60 +10,90 @@ being something like `web` or `worker`.
- * `reportStatsumErrors` - If true (the default), any errors reporting to Statsum will be reported to Sentry.
- * `process` - If set to a string that identifies this process, cpu and memory
+ * `level` - A syslog logging level. Any messages with less severity than this level will not be logged.
+ * `pretty` - Only for development use. Pretty prints logs rather than using a structured form.
```
I think we would set this via something in `main.js`/`config.yaml` via an env var there. I'll expose it in our services. Probably a good thing to do as well would be to build a cli that can turn structured logs into pretty logs.

imbstack commented on libraries/monitor/src/logger.js:
```diff
@@ -0,0 +1,144 @@
+    }
+    if (typeof fields === 'string') {
+      fields = {msg: fields};
+    }
```
My plan is to have it be used for replacing `debug` easily and the like. I will document it.

Also just wanted there to be very few cases in which this would crash.

imbstack commented on libraries/monitor/src/logger.js:
```diff
@@ -0,0 +1,144 @@
+        error: 'You may not set meta fields on logs directly.',
+        orig: fields,
+      };
+    }
```
Agreed we should change the type. I think avoiding logging the original and just logging this type will help with building dashboards from these types.

djmitche replied:
Makes sense.  It's the sort of thing we should get alerted to, just like a reportError call.

imbstack commented on libraries/monitor/src/logger.js:
```diff
@@ -0,0 +1,144 @@
+    if (fields === undefined) {
+      fields = type;
+      type = 'generic';
+    }
```
Fixed!

imbstack commented on libraries/monitor/src/index.js:
```diff
@@ -1,205 +1,235 @@
+  /*
+   * Initialize runtime dependencies that can't be preconfigured
+   * and set up process-level monitoring.
+   */
```
Yep. If you are doing this outside of the `loader` world you will have to to call setup yourself. Otherwise if it has already been setup you are ok.

imbstack commented on libraries/monitor/src/builtins.js:
```diff
@@ -0,0 +1,105 @@
+      key: 'A key that should be unique to the logger prefix.',
+      duration: 'The duration in ms of the function.',
+    },
+  },
```
I guess it isn't useful in general but I guess it doesn't hurt much either. I'm going to stick with `duration` for now although `durationMs` is probably more correct. I just have a thing against camelCase for some reason.

imbstack commented on libraries/monitor/src/builtins.js:
```diff
@@ -0,0 +1,105 @@
+      key: 'A key that should be unique to the logger prefix.',
+      duration: 'The duration in ms of whatever was timed.',
+    },
+  },
```
I fixed this by actually making it less generic. Good catch.

imbstack commented on libraries/monitor/README.md:
```diff
@@ -10,60 +10,90 @@ being something like `web` or `worker`.
+
+### Measuring and Counting Things
 
 To record a current measurement of a named value:
```
That could definitely make sense. Let's land this as-is and see how much we want that later. I've updated the readme to be a bit more specific here.

imbstack commented on libraries/monitor/src/index.js:
```diff
@@ -1,205 +1,235 @@
+    level,
+    version,
+    description,
+    fields = {}, // TODO: Consider making these defined with json-schema and validate only in dev or something
```
Ah yeah. That would be nice. Created followup bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1529472

imbstack commented:
Ok, I believe I've addressed most issues above and filed bugs for remaining ones. I've also added a potentially controversial bit (dustin and I discussed something like this without specifics during the day, what do you think, @jhford):

For compatibility with stackdriver's logging view and sentry-like feautures, if a field is called `stack` or `message`, then the top-level will have an added field called `Message` that includes this field. I leave the original in `Fields` alone to be less surprising. If neither of those fields exist, `Message` will be omitted. Alternatively we could switch to `Monitor` being `null` in these cases and always defined?

djmitche commented:
You didn't mention why that's useful :)
